src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingMiddleware.cs

```diff
@@ -60,58 +65,340 @@ public class ResponseCachingMiddleware
-                    // Can this request be served from cache?
-                    if (await cachingContext.TryServeFromCacheAsync())
+                    var response = context.HttpContext.Response;
+                    // Copy the cached status code and response headers
+                    response.StatusCode = context.CachedResponse.StatusCode;
+                    foreach (var header in context.CachedResponse.Headers)
                     {
-                        return;
+                        response.Headers.Add(header);
                     }
 
-                    // Hook up to listen to the response stream
-                    cachingContext.ShimResponseStream();
+                    response.Headers[HeaderNames.Age] = context.CachedEntryAge.TotalSeconds.ToString("F0", CultureInfo.InvariantCulture);
+
+                    var body = context.CachedResponse.Body ??
+                        ((CachedResponseBody)context.Cache.Get(context.CachedResponse.BodyKeyPrefix))?.Body;
+
+                    // If the body is not found, something went wrong.
+                    if (body == null)
+                    {
+                        return false;
+                    }
 
-                    try
+                    // Copy the cached response body
+                    if (body.Length > 0)
                     {
-                        // Subscribe to OnStarting event
-                        context.Response.OnStarting(OnStartingCallback, cachingContext);
+                        // Add a content-length if required
+                        if (response.ContentLength == null && StringValues.IsNullOrEmpty(response.Headers[HeaderNames.TransferEncoding]))
+                        {
+                            response.ContentLength = body.Length;
+                        }
+                        await response.Body.WriteAsync(body, 0, body.Length);
+                    }
+                }
 
-                        await _next(context);
+                return true;
+            }
+            else
+            {
+                // TODO: Validate with endpoint instead
+            }
 
-                        // If there was no response body, check the response headers now. We can cache things like redirects.
-                        cachingContext.OnResponseStarting();
+            return false;
+        }
 
-                        // Finalize the cache entry
-                        cachingContext.FinalizeCachingBody();
+        internal static async Task<bool> TryServeFromCacheAsync(ResponseCachingContext context)
+        {
+            foreach (var baseKey in context.KeyProvider.CreateLookupBaseKeys(context))
+            {
+                var cacheEntry = context.Cache.Get(baseKey);
+
+                if (cacheEntry is CachedVaryRules)
+                {
+                    // Request contains vary rules, recompute key(s) and try again
+                    context.CachedVaryRules = cacheEntry as CachedVaryRules;
```

Comment:
C-style cast since this is already `CachedVaryRules`. Similarly other places in this method